### PR TITLE
WIP - Initial allowlist for DB tables

### PR DESCRIPTION
**Pull request by ericqweinstein**
@wcpines Still pretty WIP, but LMKWYT! We'll probably want to rewrite the raw query with `Ecto.Query` to avoid SQLi.

**Update**: I think this is mostly good to go, and we should have basic SQLi mitigation by stripping out anything that's not `[a-zA-Z0-9_]`.

### CODE REVIEW

**Comment by ericqweinstein** [comment]
Updated `config/config.exs` in my `procore-reporting` branch:

```elixir
config :plsm,
module_name: "ProcoreReporting.BiTemporal",
destination: "lib/procore_reporting/bi_temporal/schemas/generated/",
include_tables: ["budget_forecasts", "budget_line_items"],
server: "localhost",
port: "5432",
database_name: "procore_reporting_dev",
username: "postgres",
password: "postgres",
type: :postgres
```

🎉 
```sh
ericweinstein[~/work/clients/procore-reporting] (feat/plsm *) λ mix plsm
Using PostgreSQL...
lib/procore_reporting/bi_temporal/schemas/generated/budget_line_items.ex
lib/procore_reporting/bi_temporal/schemas/generated/budget_forecasts.ex
```

**Comment by wcpines** [comment]
Per slack discussion, some ideas: 

- Change name to something like `table_filters` (e.g. `get_tables(db, table_filters)`)
- Table filters = white list or black list, specified as file name from config
- Actual filters live in newline-sep'd text file 
- If both specified, default to more restrictive whitelist

**Comment by ericqweinstein** [comment]
Will make the following changes:

1. Move from list of table names to file, something like:

```elixir
File.read(Application.get_env(:blacklist_file)))

# Meanwhile, back at the ranch

blacklist_contents |> String.split("\n", trim: true)
```

2. Add both blacklist and whitelist, where whitelist has priority if both are included

**Comment by ericqweinstein** [comment]
@wcpines Whitelist is working, blacklist still a WIP, but wanted to push up to validate the approach.

**Comment by ericqweinstein** [comment]
Progress! (**To-Do**: error handling for missing files, tests, and ensuring that whitelist takes precedent over blacklist.)

Blacklist:
```txt
budget_forecasts
budget_line_items
```

Config:
```elixir
config :plsm,
module_name: "ProcoreReporting.BiTemporal",
destination: "lib/procore_reporting/bi_temporal/schemas/generated/",
table_filters: %{exclude: "blacklist.txt"},
server: "localhost",
port: "5432",
database_name: "procore_reporting_dev",
username: "postgres",
password: "postgres",
type: :postgres
```

```sh
ericweinstein[~/work/clients/procore-reporting] (feat/plsm *) λ mix plsm
Using PostgreSQL...
lib/procore_reporting/bi_temporal/schemas/generated/schema_migrations.ex
lib/procore_reporting/bi_temporal/schemas/generated/events.ex
lib/procore_reporting/bi_temporal/schemas/generated/companies.ex
lib/procore_reporting/bi_temporal/schemas/generated/commitment_contracts.ex
lib/procore_reporting/bi_temporal/schemas/generated/projects.ex
lib/procore_reporting/bi_temporal/schemas/generated/prime_contracts.ex
lib/procore_reporting/bi_temporal/schemas/generated/prime_line_items.ex
lib/procore_reporting/bi_temporal/schemas/generated/commitment_line_items.ex
lib/procore_reporting/bi_temporal/schemas/generated/direct_costs.ex
lib/procore_reporting/bi_temporal/schemas/generated/line_item_types.ex
lib/procore_reporting/bi_temporal/schemas/generated/sub_jobs.ex
lib/procore_reporting/bi_temporal/schemas/generated/cost_codes.ex
lib/procore_reporting/bi_temporal/schemas/generated/wbs_codes.ex
lib/procore_reporting/bi_temporal/schemas/generated/direct_cost_line_items.ex
lib/procore_reporting/bi_temporal/schemas/generated/potential_change_orders.ex
lib/procore_reporting/bi_temporal/schemas/generated/markups.ex
lib/procore_reporting/bi_temporal/schemas/generated/change_order_packages.ex
lib/procore_reporting/bi_temporal/schemas/generated/change_order_requests.ex
lib/procore_reporting/bi_temporal/schemas/generated/project_configurations.ex
lib/procore_reporting/bi_temporal/schemas/generated/markup_line_items.ex
lib/procore_reporting/bi_temporal/schemas/generated/budget_line_item_transfers.ex
lib/procore_reporting/bi_temporal/schemas/generated/monitoring_resources.ex
lib/procore_reporting/bi_temporal/schemas/generated/sub_invoice_items.ex
```

Whitelist:
```txt
budget_forecasts
budget_line_items
```

Config:
```elixir
config :plsm,
module_name: "ProcoreReporting.BiTemporal",
destination: "lib/procore_reporting/bi_temporal/schemas/generated/",
table_filters: %{include: "whitelist.txt"},
server: "localhost",
port: "5432",
database_name: "procore_reporting_dev",
username: "postgres",
password: "postgres",
type: :postgres
```

```sh
ericweinstein[~/work/clients/procore-reporting] (feat/plsm *) λ mix plsm
Using PostgreSQL...
lib/procore_reporting/bi_temporal/schemas/generated/budget_line_items.ex
lib/procore_reporting/bi_temporal/schemas/generated/budget_forecasts.ex
```

**Comment by wcpines** [comment]
Thoughts on renaming anything that isn't the actual `config/config*` to `configuration`?

**Comment by wcpines** [comment]
Thoughts on renaming non config files (i.e. anything actually under `config/` to `configuration` 

 E.g. 

`defmodule Plsm.Configuration.ConfigWriter do...`

`def write(file_name) do`  => `def run(file_name) do`

File: lib/database/postgresql.ex [hunk 1]
```diff
@@ -142,4 +164,19 @@ defimpl Plsm.Database, for: Plsm.Database.PostgreSQL do
       true -> :none
     end
   end
+
+  defp list_to_sql(table_names) do
+    table_names
```
Comment (wcpines) [inline lib/database/postgresql.ex]: @ericqweinstein Maybe we can be a little extra safe here and add a step to strip any non alphanumeric and `_` chars? Not sure if needed.

File: lib/database/postgresql.ex [hunk 2]
```diff
@@ -40,19 +40,39 @@ defimpl Plsm.Database, for: Plsm.Database.PostgreSQL do
...
-    |> Enum.map(fn x -> %Plsm.Database.TableHeader{database: db, name: x} end)
+  @spec get_tables(Plsm.Database.PostgreSQL, %{
+          optional(:include) => String.t(),
+          optional(:exclude) => String.t()
+        }) :: [Plsm.Database.TableHeader]
+  def get_tables(db, %{include: whitelist} = table_filters) do
+    Postgrex.query!(
+      db.connection,
+      "SELECT table_name FROM information_schema.tables WHERE table_schema = 'public' AND table_name IN $1;",
+      list_to_sql(whitelist)
```
Comment (wcpines) [inline lib/database/postgresql.ex]: Haven't completed writing tests for this implementation yet. I'm curious if this is sufficient to sanitize inputs (i.e. using numbered args)

File: lib/database/postgresql.ex [hunk 3]
```diff
@@ -49,12 +49,10 @@ defimpl Plsm.Database, for: Plsm.Database.PostgreSQL do
...
   def get_tables(db, %{include: whitelist} = table_filters) do
-    Postgrex.query!(
-      db.connection,
-      "SELECT table_name FROM information_schema.tables WHERE table_schema = 'public' AND table_name IN $1;",
-      list_to_sql(whitelist)
-    )
-    |> handle_table_results(db)
+    whitelist
+    |> Enum.map(fn table_name ->
+      %Plsm.Database.TableHeader{database: db, name: table_name}
```
Comment (ericqweinstein) [inline lib/database/postgresql.ex]: 💯

File: lib/database/postgresql.ex (hunk 1 above)
Comment (ericqweinstein) [reply lib/database/postgresql.ex]: Should be easy enough! I think we can just match on `\W+` then:

```elixir
String.replace("Robert'; DROP TABLE Students;--", ~r/\W+/, "", global: true)
# => "RobertDROPTABLEStudents"
```

File: lib/database/postgresql.ex [hunk 4]
```diff
@@ -162,20 +190,4 @@ defimpl Plsm.Database, for: Plsm.Database.PostgreSQL do
       true -> :none
     end
   end
-
-  def list_to_sql(table_names) do
```
Comment (ericqweinstein) [inline lib/database/postgresql.ex]: Just moved these up—not sure the Elixir convention, but coming from Rubyland, I expect to see private functions at the bottom of the file. Happy to move 'em back, though!

File: test/database/postgresql_test.exs [hunk 5]
```diff
@@ -13,30 +13,30 @@ defmodule Plsm.Database.PostgreSQLTest do
...
-                 port: _port,
-                 server: _server,
-                 username: _user
-               },
-               name: name
-             }
-
-             name not in bl
-           end)
+    # configs = Plsm.Common.Configs.load()
```
Comment (ericqweinstein) [inline test/database/postgresql_test.exs]: Temporarily commented out while writing the above tests, will restore shortly 👍

File: test/database/common_test.exs [hunk 6]
```diff
@@ -0,0 +1,9 @@
+defmodule Plsm.Database.CommonTest do
+  alias Plsm.Database.Common
+  use ExUnit.Case
+
+  test("list_to_sql/1") do
+    sanitized_table_names = Plsm.Database.Common.list_to_sql(["Robert'; DROP TABLE Students;--"])
+    assert sanitized_table_names == ["'RobertDROPTABLEStudents'"]
```
Comment (ericqweinstein) [inline test/database/common_test.exs]: We can and should write better/more comprehensive tests here, just wanted to get the ball rolling

**Comment by ericqweinstein** [comment]
We'll need to add new tests, but the existing ones are now passing:

```sh
λ mix test
......

Finished in 0.05 seconds
6 tests, 0 failures

Randomized with seed 622606
```

File: lib/database/postgresql.ex (hunk 3 above)
Comment (ericqweinstein) [reply lib/database/postgresql.ex]: I'm still wrapping my head around multiple function heads in Elixir, but it seems to me these definitions don't depend on each other, so I'm not 100% sure why this works without an explicit `Postgrex.query!` invocation

Comment (ericqweinstein) [reply lib/database/postgresql.ex]: NVM, was just confused that we pass the connection but don't need to use it 👍

File: test/database/common_test.exs (hunk 6 above)
Comment (wcpines) [reply test/database/common_test.exs]: nah ;)

**Comment by ericqweinstein** [comment]
As mentioned, we may want to implement `get_tables/2` for MySQL, but otherwise lookin' good:

```
λ mix test
Compiling 10 files (.ex)
warning: function get_tables/2 required by protocol Plsm.Database is not implemented (in module Plsm.Database.Plsm.Database.MySql)
  lib/database/mysql.ex:10

Generated plsm app
........

Finished in 0.04 seconds
8 tests, 0 failures

Randomized with seed 536055
```

File: lib/database/mysql.ex [hunk 7]
```diff
@@ -44,14 +44,37 @@ defimpl Plsm.Database, for: Plsm.Database.MySql do
...
+    |> Enum.map(fn table_name ->
+      %Plsm.Database.TableHeader{database: db, name: table_name}
+    end)
+  end
 
-    result.rows
-    |> List.flatten()
-    |> Enum.map(fn x -> %Plsm.Database.TableHeader{database: db, name: x} end)
+  def get_tables(db, %{exclude: blacklist} = _table_filters) do
+    Mariaex.query!(
```
Comment (ericqweinstein) [inline lib/database/mysql.ex]: @wcpines Reading the docs on this now to double check this will work; setting up a little MySQL locally to test

Comment (ericqweinstein) [reply lib/database/mysql.ex]: @wcpines Confirming that this all works correctly for MySQL:

```
λ mix plsm
Using MySql...
lib/procore_reporting/bi_temporal/schemas/generated/test_table_1.ex
lib/procore_reporting/bi_temporal/schemas/generated/test_table_2.ex
```

I couldn't get this to work with MySQL 8, though; I had to back down to 5.7. Might be worth our adding to the README.

**Comment by ericqweinstein** [comment]
@wcpines Lookin' good to merge? LMK if other changes needed (as mentioned, we might want to mention that Plsm doesn't work with latest MySQL).

File: README.md [hunk 8]
```diff
@@ -1,50 +1,75 @@
...
+    {:plsm, "~> 2.2.0"},
+  # ...
+  ]
+end
+```
 
-First, in order to run plsm, you need to generate a config file. You do this by running
+### Running Plsm
+
+In order to run plsm, first generate a config file (if you are using Phoenix, it is recommended to to pass in your `config/dev.exs`, as Plsm is a dev-only process):
```
Comment (wcpines) [inline README.md]: Capitalize non-code/cmd references to Plsm

File: README.md [hunk 9]
```diff
@@ -1,50 +1,75 @@
...
-Once you have your config file generated then you are ready to run plsm. You do this by running 
+This will create a skeleton config in `<name>`. If none is specified, the config is appended to your `config/config.exs`.
+
+
+You will also need to make changes to the generated Plsm configs in the config file in order to allow Plsm to function correctly.
+
+
+#### Options
+
+You can blacklist or whitelist database tables using the `:tables_filters` option in your config.
```
Comment (wcpines) [inline README.md]: remove `s` in `table_filters`

**Pull request merged**